### PR TITLE
Debouncing

### PR DESCRIPTION
**Pull request by IdoPeled8**
add debounce for test suits.
fix test components.
move state to store.
cleaner code & less code

### CODE REVIEW

File: client/src/pages/test-suites/RequestModal.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import { useState } from "react";
+import TestButtonSection from "./TestButtonSection";
+
+interface RequestModalProps {
+  onTestChange: (test: TestType) => void;
+  onTestMethodChange: (test: TestType) => void;
+  onRequestHeadersChange: (rules: any[]) => void;
+  requestHeaders: any[];
+}
+const RequestModal = ({
```
Comment (ypotter) [inline client/src/pages/test-suites/RequestModal.tsx:16]: This is not a modal

File: client/src/pages/test-suites/TestSuiteMainSection.tsx [hunk 2]
```diff
@@ -1,153 +1,231 @@
...
+    console.log("save all");
+    if (!testSuiteId || !testId || !currentTest) {
+      return;
+    }
+    saveTest(testSuiteId, testId, {
+      ...currentTest,
+      rules: [statusCodeRule, bodyRule, ...headerRules],
     });
-    setHeaderRules(headers);
+    if (!testSuiteId || !testId || !currentTest) {
```
Comment (ypotter) [inline client/src/pages/test-suites/TestSuiteMainSection.tsx:160]: duplicated

File: client/src/pages/test-suites/TestSuiteMainSection.tsx [hunk 3]
```diff
@@ -1,153 +1,231 @@
...
+      return;
+    }
+  }, [currentTest, testSuiteId, testId, statusCodeRule, bodyRule, headerRules]);
+
+  const saveTest = (
```
Comment (ypotter) [inline client/src/pages/test-suites/TestSuiteMainSection.tsx:165]: Couldn't it be in the testStore?

File: client/src/pages/test-suites/TestSuiteMainSection.tsx [hunk 4]
```diff
@@ -1,153 +1,231 @@
...
 
-  const extractHeaders = (test: TestType) => {
-    const headers: Rule[] = [];
-    test.rules.forEach((rule) => {
-      if (rule.type === "header") {
-        headers.push(rule);
-      }
+  const handleSaveAll = React.useCallback(() => {
+    console.log("save all");
+    if (!testSuiteId || !testId || !currentTest) {
```
Comment (ypotter) [inline client/src/pages/test-suites/TestSuiteMainSection.tsx:153]: This validation could be in saveTest and save a lot of duplicated code

File: client/src/pages/test-suites/TestSuiteMainSection.tsx [hunk 5]
```diff
@@ -1,153 +1,231 @@
...
-import { Tooltip } from "@mui/material";
 import { TestConfig } from "./TestConfig";
-import { TbAdjustmentsCog } from "react-icons/tb";
 import { ExecutionHistory } from "./ExecutionHistory";
+import TestsTopSection from "./TestsTopSection";
+import TabContext from "@mui/lab/TabContext";
+import RequestModal from "./RequestModal";
+import AssertionsModal from "./AssertionsModal";
 
 export const TestSuiteMainSection = () => {
```
Comment (ypotter) [inline client/src/pages/test-suites/TestSuiteMainSection.tsx:12]: The functions here are repetitive with small changes between them. generally it confuses to read this code and understand the flow. I would this of a way to simplify it.

**Comment by IdoPeled8** [comment]
image

**Review by ypotter — CHANGES_REQUESTED** [review]

**Pull request merged**